File: examples/02-rest-tour/ports/go/main.go
// Workshop REST tour — Pillar 2 (Go).
//
// Pure net/http. 4 demo endpoints exercising rest.RestClient.

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/signalwire/signalwire-go/pkg/rest"
)

type creds struct{ ProjectID, Space, Token string }

var (
	mu sync.Mutex
	c  *creds
)

func client() (*rest.RestClient, error) {
	mu.Lock()
	cc := c
	mu.Unlock()
	if cc == nil {
		return nil, fmt.Errorf("run setup first")
	}
	return rest.NewRestClient(cc.ProjectID, cc.Token, cc.Space)
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func sharedUIDir() string {
	wd, _ := os.Getwd()
	return filepath.Join(wd, "..", "..", "..", "..", "shared", "ui")
}

func main() {
	port := os.Getenv("PORT")
	if port == "" {
		port = "8001"
	}
	sharedDir := sharedUIDir()

	mux := http.NewServeMux()
	mux.Handle("/shared/", http.StripPrefix("/shared/", http.FileServer(http.Dir(sharedDir))))

	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/" {
			http.NotFound(w, r)
			return
		}
		http.ServeFile(w, r, filepath.Join(sharedDir, "creds-form.html"))
	})

	mux.HandleFunc("/demo.js", func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/javascript")
		http.ServeFile(w, r, "demo.js")
	})

	mux.HandleFunc("/api/setup", func(w http.ResponseWriter, r *http.Request) {
		var body struct {
			ProjectID string `json:"project_id"`
			Space     string `json:"space"`
			Token     string `json:"token"`
		}
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "invalid json"})
			return
		}
		pid := strings.TrimSpace(body.ProjectID)
		sp := strings.TrimSpace(body.Space)
		tk := strings.TrimSpace(body.Token)
		if pid == "" || sp == "" || tk == "" {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "All fields required"})
			return
		}
		rc, err := rest.NewRestClient(pid, tk, sp)
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		if _, err := rc.PhoneNumbers.List(map[string]any{"limit": 1}); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": fmt.Sprintf("Credential check failed: %v", err)})
			return
		}
		mu.Lock()
		c = &creds{ProjectID: pid, Space: sp, Token: tk}
		mu.Unlock()
		writeJSON(w, 200, map[string]any{"ok": true, "jwt": "session-validated", "subscriber_id": "n/a"})
	})

	mux.HandleFunc("/api/list-numbers", func(w http.ResponseWriter, r *http.Request) {
		rc, err := client()
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		resp, err := rc.PhoneNumbers.List(map[string]any{"limit": 20})
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok":       true,
			"sdk_call": `client.PhoneNumbers.List(map[string]any{"limit": 20})`,
			"response": resp,
		})
	})

	mux.HandleFunc("/api/send-sms", func(w http.ResponseWriter, r *http.Request) {
		rc, err := client()
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		var body struct{ From, To, Body string }
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "invalid json"})
			return
		}
		from := strings.TrimSpace(body.From)
		to := strings.TrimSpace(body.To)
		text := strings.TrimSpace(body.Body)
		if text == "" {
			text = "Hello from the SignalWire workshop!"
		}
		if from == "" || to == "" {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "from + to required"})
			return
		}
		resp, err := rc.Compat.Messages.Create(map[string]any{"from": from, "to": to, "body": text})
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok":       true,
			"sdk_call": fmt.Sprintf(`client.Compat.Messages.Create(...from: %q, to: %q...)`, from, to),
			"response": resp,
		})
	})

	mux.HandleFunc("/api/recent-calls", func(w http.ResponseWriter, r *http.Request) {
		rc, err := client()
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		resp, err := rc.Compat.Calls.List(map[string]any{"page_size": 10})
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok":       true,
			"sdk_call": `client.Compat.Calls.List(map[string]any{"page_size": 10})`,
			"response": resp,
		})
	})

	mux.HandleFunc("/api/wire-number", func(w http.ResponseWriter, r *http.Request) {
		rc, err := client()
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		var body struct{ SID, VoiceURL string }
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "invalid json"})
			return
		}
		sid := strings.TrimSpace(body.SID)
		voiceURL := strings.TrimSpace(body.VoiceURL)
		if sid == "" || voiceURL == "" {
			writeJSON(w, 400, map[string]any{"ok": false, "error": "sid + voice_url required"})
			return
		}
		resp, err := rc.PhoneNumbers.Update(sid, map[string]any{"voice_url": voiceURL, "voice_method": "POST"})
		if err != nil {
			writeJSON(w, 400, map[string]any{"ok": false, "error": err.Error()})
			return
		}
		writeJSON(w, 200, map[string]any{
			"ok":       true,
			"sdk_call": fmt.Sprintf(`client.PhoneNumbers.Update(%q, voice_url=%q)`, sid, voiceURL),
			"response": resp,
		})
	})

	addr := "0.0.0.0:" + port
	log.Printf("REST tour listening on %s", addr)
	if err := http.ListenAndServe(addr, mux); err != nil {
		log.Fatal(err)
	}
}

File: examples/02-rest-tour/ports/rust/src/main.rs
//! Workshop REST tour — Pillar 2 (Rust).
//!
//! Axum + signalwire RestClient. 4 demo endpoints calling phone_numbers,
//! compat.messages, compat.calls, phone_numbers update.

use axum::{
    extract::State,
    http::StatusCode,
    response::{IntoResponse, Json},
    routing::{get, post},
    Router,
};
use serde::Deserialize;
use serde_json::{json, Value};
use signalwire::rest::client::RestClient;
use std::{net::SocketAddr, path::PathBuf, sync::Mutex};
use tokio::sync::OnceCell;
use tower_http::services::ServeDir;

#[derive(Default)]
struct AppState {
    creds: Mutex<Option<(String, String, String)>>,
}

static STATE: OnceCell<AppState> = OnceCell::const_new();

fn shared_ui_dir() -> PathBuf {
    let cwd = std::env::current_dir().unwrap();
    cwd.join("..").join("..").join("..").join("..").join("shared").join("ui").canonicalize()
        .unwrap_or_else(|_| cwd.join("../../../../shared/ui"))
}

fn client() -> Result<RestClient, String> {
    let g = STATE.get().unwrap().creds.lock().unwrap();
    let (p, sp, tk) = g.as_ref().ok_or("Run setup first")?.clone();
    drop(g);
    Ok(RestClient::new(&p, &tk, &sp))
}

#[derive(Deserialize)]
struct SetupBody { project_id: String, space: String, token: String }

#[derive(Deserialize)]
struct SmsBody { from: String, to: String, body: Option<String> }

#[derive(Deserialize)]
struct WireBody { sid: String, voice_url: String }

async fn setup(Json(b): Json<SetupBody>) -> impl IntoResponse {
    let pid = b.project_id.trim();
    let sp = b.space.trim();
    let tk = b.token.trim();
    if pid.is_empty() || sp.is_empty() || tk.is_empty() {
        return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": "All fields required"})));
    }
    let c = RestClient::new(pid, tk, sp);
    // Validate by listing numbers (best-effort)
    if let Err(e) = c.phone_numbers().list(json!({"limit": 1})).await {
        return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": format!("Credential check failed: {}", e)})));
    }
    *STATE.get().unwrap().creds.lock().unwrap() = Some((pid.into(), sp.into(), tk.into()));
    (StatusCode::OK, Json(json!({"ok": true, "jwt": "session-validated", "subscriber_id": "n/a"})))
}

async fn list_numbers() -> impl IntoResponse {
    let c = match client() { Ok(c) => c, Err(e) => return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e}))) };
    match c.phone_numbers().list(json!({"limit": 20})).await {
        Ok(r) => (StatusCode::OK, Json(json!({"ok": true, "sdk_call": "client.phone_numbers().list({limit: 20})", "response": r}))),
        Err(e) => (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e.to_string()}))),
    }
}

async fn send_sms(Json(b): Json<SmsBody>) -> impl IntoResponse {
    let c = match client() { Ok(c) => c, Err(e) => return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e}))) };
    let body = b.body.unwrap_or_else(|| "Hello from the SignalWire workshop!".into());
    let payload = json!({"from": &b.from, "to": &b.to, "body": &body});
    match c.compat().messages().create(payload).await {
        Ok(r) => (StatusCode::OK, Json(json!({"ok": true, "sdk_call": format!("client.compat().messages().create(from: {}, to: {})", b.from, b.to), "response": r}))),
        Err(e) => (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e.to_string()}))),
    }
}

async fn recent_calls() -> impl IntoResponse {
    let c = match client() { Ok(c) => c, Err(e) => return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e}))) };
    match c.compat().calls().list(json!({"page_size": 10})).await {
        Ok(r) => (StatusCode::OK, Json(json!({"ok": true, "sdk_call": "client.compat().calls().list({page_size: 10})", "response": r}))),
        Err(e) => (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e.to_string()}))),
    }
}

async fn wire_number(Json(b): Json<WireBody>) -> impl IntoResponse {
    let c = match client() { Ok(c) => c, Err(e) => return (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e}))) };
    match c.phone_numbers().update(&b.sid, json!({"voice_url": &b.voice_url, "voice_method": "POST"})).await {
        Ok(r) => (StatusCode::OK, Json(json!({"ok": true, "sdk_call": format!("client.phone_numbers().update({}, voice_url: {})", b.sid, b.voice_url), "response": r}))),
        Err(e) => (StatusCode::BAD_REQUEST, Json(json!({"ok": false, "error": e.to_string()}))),
    }
}

#[tokio::main]
async fn main() {
    STATE.set(AppState::default()).ok();
    let shared = shared_ui_dir();
    let app = Router::new()
        .route("/api/setup", post(setup))
        .route("/api/list-numbers", get(list_numbers))
        .route("/api/send-sms", post(send_sms))
        .route("/api/recent-calls", get(recent_calls))
        .route("/api/wire-number", post(wire_number))
        .nest_service("/shared", ServeDir::new(&shared))
        .route("/", get(move || {
            let path = shared.join("creds-form.html");
            async move { (
                [(axum::http::header::CONTENT_TYPE, "text/html")],
                std::fs::read_to_string(&path).unwrap_or_default(),
            ) }
        }))
        .route("/demo.js", get(|| async {
            (
                [(axum::http::header::CONTENT_TYPE, "application/javascript")],
                std::fs::read_to_string("demo.js").unwrap_or_default(),
            )
        }));

    let port: u16 = std::env::var("PORT").ok().and_then(|s| s.parse().ok()).unwrap_or(8001);
    let addr: SocketAddr = ([0, 0, 0, 0], port).into();
    println!("REST tour listening on http://{}", addr);
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
